src/kindtech/geo/_arcgis/_ingestion.py

```diff
@@ -0,0 +1,185 @@
+        resolution, raw name)
+    """
+    logger.info("Parsing HTML for services")
+    kept = []
+
+    soup = BeautifulSoup(html_content, "html.parser")
+    services_list = soup.find("ul", id="servicesList")
+
+    if not services_list:
+        logger.warning("No services list found in HTML content")
+        return kept
+
+    for li in services_list.find_all("li"):
+        link = li.find("a")
+        if link and "FeatureServer" in link["href"]:
+            service_name = link.text
+
+            kept.append(service_name)
+
+    logger.info(f"Total FeatureServer services found: {len(kept)}")
+
+    return kept
+
+
+def _process_and_save(services: list):
+    df = pl.DataFrame({"arcgis_id": services})
+
+    print(df.height)
+
+    to_skip = [
+        "LAD_Dec_1961_in_England_and_Wales_BFC_Boundaries_2022",
+        "LAD_JUN_1921_EW_BGC",
+        "LAD_PT_JUN_1921_EW_BGC",
+        "LAD_DEC_2021_EW_BFE_RUC",
+        "LAD_DEC_2024_EW_BFE_RUC",
+    ]
+
+    # case-insensitive, allow “noise” tokens between the key bits
+    pattern = r"""(?ix)                             # i = ignore-case, x = verbose
+    ^(?:LAD|LOCAL_AUTHORITY_DISTRICTS).*?_
+    (?:                                             # ── optional month ────────────
+    (?P<month>JAN(?:UARY)?|FEB(?:RUARY)?|MAR(?:CH)?|APR(?:IL)?|
+                MAY|JUN(?:E)?|JUL(?:Y)?|AUG(?:UST)?|
+                SEP(?:TEMBER)?|OCT(?:OBER)?|NOV(?:EMBER)?|DEC(?:EMBER)?)
+    _
+    )?                                              # ──────────────────────────────
+    (?P<year>\d{2}(?:\d{2})?)_                      # 24  or  2024, 1961, …
+    (?:[A-Z]+_)*?                                   # skip words like BOUNDARIES
+    (?P<region>UK|GB|EW)_
+    (?:[A-Z]+_)*?                                   # more noise
+    (?P<res>BFC|BFE|BGC|BSC)                        # resolution code
+    (?:_.*)?$                                       # allow V3_2022 etc. at the end
+    """
+
+    df = df.filter(
+        (pl.col("arcgis_id").str.contains(pattern))
+        & (~pl.col("arcgis_id").is_in(to_skip))
+    ).with_columns(
+        [
+            pl.lit("LAD").alias("geography"),
+            pl.col("arcgis_id")
+            .str.extract(pattern, group_index=2)
+            .cast(pl.Int16)
+            .alias("year"),
+            pl.col("arcgis_id").str.extract(pattern, group_index=1).alias("month"),
+            pl.col("arcgis_id").str.extract(pattern, group_index=3).alias("region"),
+            pl.col("arcgis_id").str.extract(pattern, group_index=4).alias("resolution"),
+        ]
+    )
+
+    df.write_csv(DEFAULT_OUTPUT_PATH)
+
+    print(df.height)
+    return df.height
+
+
+def ingest_arcgis_services(url: str = ARCGIS_BASE_URL) -> int:
+    """
+    Main function to ingest ArcGIS services from the web and save them to a CSV file.
+
+    Args:
+        url: The URL of the ArcGIS REST Services catalog.
+
+    Returns:
+        The number of services ingested.
```

Comment:
Should we return the service number or just output it as a logging item? Instead we can return the paths for the argis files perhaps?

Reply:
Sounds good too, I'll change that